### PR TITLE
Apps v2 service checks

### PR DESCRIPTION
**Pull request by dangra**
Respect checks defined for services so fly-proxy can relate them and use their status to decide when to route traffic to instances serving that service or not.

### CODE REVIEW

File: internal/app/app.go [hunk 1]
```diff
@@ -637,45 +655,31 @@ func (c *Config) SetVolumes(volumes []scanner.Volume) {
...
-	defaultProcessName := ""
-	firstProcessNameOrDefault := ""
-	if processCount == 1 {
-		for procName := range c.Processes {
-			firstProcessNameOrDefault = procName
-			break
-		}
+	processCount := len(c.Processes)
+	if processCount == 0 {
+		c.Processes = map[string]string{"": ""}
```
Comment (tvdfly) [inline internal/app/app.go:667]: the original was causing a panic when c.Processes was nil, and this version fixes that. Setting c.Processes here does break assumptions elsewhere that empty processes means no processes were set.  i'll try not overwriting c.Processes and see if that fixes stuff...

Might be good to do some more testing before merging this.

File: internal/app/app.go [hunk 2]
```diff
@@ -133,26 +134,33 @@ func (hs *HttpService) ToMachineService() *api.MachineService {
 		}},
 		Concurrency: concurrency,
 	}
-
 }
 
-func (s *Service) ToMachineService() *api.MachineService {
+func (s *Service) toMachineService() *api.MachineService {
+	checks := make([]api.Check, len(s.TCPChecks)+len(s.HTTPChecks))
```
Comment (tvdfly) [inline internal/app/app.go:140]: I'm pretty sure this makes a slice with `len(s.TCPChecks)+len(s.HTTPChecks)` items all set to nil. the appends add to that list without overwriting any of the nils. 

```suggestion
	checks := make([]api.Check, 0)
```

Comment (tvdfly) [reply internal/app/app.go:140]: I'm playing around with this, and will push an update in a sec.

**Review by tvdfly — COMMENTED** [review]
it looks like the per service check is getting passed along to the machines, and flaps doesn't yet support those checks (which I'm thinking is expected and the flaps change will come next).

setting the top level checks in fly.toml still works. it looks like we're not letting it use the newly supported attributes (like grace period). Is it safe to add those?

**Comment by tvdfly** [comment]
btw, thank you for cleaning up this code and making it simpler! It's much easier to go through now.

File: internal/app/app.go [hunk 3]
```diff
@@ -178,7 +186,7 @@ type Check struct {
 	Headers       map[string]string `json:"headers,omitempty" toml:"headers,omitempty"`
 }
 
-func (c *Check) toMachineCheck(launching bool) (*api.MachineCheck, error) {
+func (c *ToplevelCheck) toMachineCheck(launching bool) (*api.MachineCheck, error) {
```
[the remark this answers is not in the document]
Comment (dangra) [reply internal/app/app.go:189]: Yup, I was waiting on the server side changes to be ready before pushing it. Will do soon.

File: internal/app/app.go (hunk 1 above)
Comment (dangra) [reply internal/app/app.go]: thanks for this!

File: internal/app/app.go [hunk 4]
```diff
@@ -137,7 +137,7 @@ func (hs *HTTPService) toMachineService() *api.MachineService {
 }
 
 func (s *Service) toMachineService() *api.MachineService {
-	checks := make([]api.Check, len(s.TCPChecks)+len(s.HTTPChecks))
+	checks := make([]api.Check, 0)
```
Comment (dangra) [inline internal/app/app.go:140]: Oh, my mistake, the intention was to create an empty slice with the expected capacity

```
checks := make([]api.Check, 0, len(s.TCPChecks)+len(s.HTTPChecks))
```

Comment (tvdfly) [reply internal/app/app.go:140]: Oh that makes sense. Thanks!

File: internal/app/app.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (tvdfly) [reply internal/app/app.go:189]: Awesome, thanks!

**Pull request merged**